### ISSUE TITLE
Make CI cover all public XML_ build features

### ISSUE THREAD

**Issue by hartwork** [libexpat/libexpat#324]
Not all public build flags are covered — both enabled and disabled — by CI yet. This is what's covered:
- [x] `XML_ATTR_INFO` (default: OFF) —  #273
- [ ] `XML_CONTEXT_BYTES` (default: ON, 1024 bytes)
- [ ] `XML_DTD` (default: ON)
- [ ] `XML_LARGE_SIZE` (default: OFF)
- [ ] `XML_MIN_SIZE` (default: OFF)
- [ ] `XML_NS` (default: ON)
- [x] `XML_UNICODE` (default: OFF)
- [x] `XML_UNICODE_WCHAR_T` (default: OFF)

They can be seen together here:
https://github.com/libexpat/libexpat/blob/80d2829fa80f41a7c6308d1fbc98d7310f064a03/expat/lib/xmlparse.c#L2360-L2384

It would be cool to have one isolated run on Linux (not macOS) in Travis in the include matrix … https://github.com/libexpat/libexpat/blob/a16ca4c913be156112f9c4a0d8b2a66ffa2b7f4d/.travis.yml#L22-L23 … and to fix everything needed to get there. `make run-xmltest` might need to be excluded for some, `make check` should be made to work for all.

**Comment by bettermanzzy** [comment]
Hi, I may do something for it.
 `XML_ATTR_INFO`, `XML_UNICODE` and `XML_UNICODE_WCHAR_T` features  are covered by CI. And  `XML_CONTEXT_BYTES `, `XML_DTD` and `XML_NS` features  are support by default in CI. 
So, only `XML_LARGE_SIZE` and `XML_MIN_SIZE` are not covered by CI yet. If we enable XML_LARGE_SIZE and XML_MIN_SIZE in build an CI. Make CI can cover all public XML_ build features.
@hartwork Can I understand this requirement like this?

**Comment by hartwork** [comment]
> Hi, I may do something for it.

Cool!

> And `XML_CONTEXT_BYTES `, `XML_DTD` and `XML_NS` features are support by default in CI.
> So, only `XML_LARGE_SIZE` and `XML_MIN_SIZE` are not covered by CI yet. If we enable XML_LARGE_SIZE and XML_MIN_SIZE in build an CI. Make CI can cover all public XML_ build features.
> @hartwork Can I understand this requirement like this?

For `XML_CONTEXT_BYTES `, `XML_DTD` and `XML_NS` we don't cover their disabled versions, yet. In the current CI, they are enabled everywhere. So for these three we need to cover them disabled as well.

**Comment by hartwork** [comment]
PS: In case of CMake, #326 and #328 might be of interest.

**Comment by bettermanzzy** [comment]
I am trying to do it  by add build flags in `.travis.yml` file， but CI  fails in my fork libexpat repository.
I have add follows code in `./travis.yml`
``` YAML
- os: linux
  env: MODE=cmake-oos CMAKE_ARGS="-DEXPAT_ATTR_INFO=ON -DEXPAT_CONTEXT_BYTES=1024 -DEXPAT_DTD=ON -DEXPAT_LARGE_SIZE=ON -DEXPAT_MIN_SIZE=ON -DEXPAT_NS=ON"
```
I think CI failure mainly happened in ` make VERBOSE=1 all test` according to error  meaages.  
I am confused. Could you give me some help? Thanks @hartwork

**Comment by hartwork** [comment]
I found your build at https://travis-ci.org/bettermanzzy/libexpat/builds/579137263 . The segfault at https://travis-ci.org/bettermanzzy/libexpat/jobs/579137274 may uncover a bug in the library or in test code. That's exactly why we need coverage for those.
I should mention that we should activate [AddressSanitizer](https://clang.llvm.org/docs/AddressSanitizer.html) as well — either "from the outside" in `.travis.yml` (see notes regarding `EXPAT_BUILD_FUZZERS`) or with a new option `EXPAT_ASAN` — so that all memory access bugs become fatal and produce reports.

**Comment by hartwork** [comment]
@bettermanzzy quick update: The segfault is from `XML_MIN_SIZE` in general, not `XML_LARGE_SIZE`, made a new issue #332 now. Also made #331 for the compile warnings being trouble to `qa.sh`.

If you could let me know on or offline, what parts you want to focus on that would be great so that I don't work on the same things. Thank you!

**Comment by bettermanzzy** [comment]
@hartwork thanks for your reply and tips.  `XML_MIN_SIZE` made issue #332.  I think it's a stack overflow bug, but it's hard to find the problem in code. I may continue to try to fix.

I have tried to cover  all public XML_ build features except `XML_MIN_SIZE`. The CI running result is OK (https://travis-ci.org/bettermanzzy/libexpat/builds/579703752).   
If we cannot fix issue #332 quickly, whether to cover other public XML_build features first.

**Comment by hartwork** [comment]
@bettermanzzy I have created #340 to allow us using `-DEXPAT_MIN_SIZE=ON` in the CI short term.

Once #336 is merged, the default CMake config will be equivalent to the combination of:
- `-DEXPAT_ATTR_INFO=OFF`
- `-DEXPAT_CONTEXT_BYTES=1024`
- `-DEXPAT_DTD=ON`
- `-DEXPAT_LARGE_SIZE=OFF`
- `-DEXPAT_MIN_SIZE=OFF`
- `-DEXPAT_NS=ON`
- `-DEXPAT_CHAR_TYPE=char`

I believe we mostly need one case per non-default case in isolation:
- `-DEXPAT_CONTEXT_BYTES=OFF`
- `-DEXPAT_DTD=OFF`
- `-DEXPAT_LARGE_SIZE=ON`
- `-DEXPAT_MIN_SIZE=ON`
- `-DEXPAT_NS=OFF`

Out of these, the change at https://github.com/bettermanzzy/libexpat/commit/91b5b1a565190470225970eb7707427becd9124b only covers `-DEXPAT_LARGE_SIZE=ON` (and repeats default settings without strong need to) at the moment.

**Comment by hartwork** [comment]
News pull request at #354 now
